rfd/0065-distributed-tracing.md

```diff
@@ -0,0 +1,247 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 65 - Distributed Tracing
+
+## What
+
+Add distributed tracing to Teleport by leveraging [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go)
+and its various libraries to create spans and export them in the
+[OpenTelemetry Protocol (OTLP)](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md).
+
+## Why
+
+We don't currently have much insight into areas of latency within Teleport. Consider the following questions:
+
+> _"Why is it taking so long to run `tsh ls`?"_
+>
+> _"Why does it take N seconds to connect via `tsh ssh`?"_
+
+Today we can only guess what the cause is by trying to find logs that suggest where the latency is coming from. Not all
+latency issues are straightforward to reproduce either. Some issues may be specific to the environment, or a
+deployment strategy, which we may not be able to easily/reliably reproduce. At the time of this
+writing there are a few open issues related to latency that have yet to be addressed:
+[#11159](https://github.com/gravitational/teleport/issues/11159),
+[#8149](https://github.com/gravitational/teleport/issues/8145),
+[#7815](https://github.com/gravitational/teleport/issues/7815).
+
+
+Distributed tracing provides full visibility into the life of a request across all the different service
+boundaries it may pass. By adding distributed tracing to Teleport we would be able to see precisely what is occurring
+and for how long across all the different Teleport services. With this level of observability we would gain the ability
+to detect and identify areas of concern before a customer finds them. It would also allow us to capture and observe
+spans from the customers' environment and reduce the burden we currently face to try and reproduce the latency issues.
+
+## Details
+
+### Goals
+
+* Generate and export spans for Teleport service boundaries and core components
+  (e.g. http client/server, grpc client/server, ssh client/server, cache, backend)
+* Provide a solid foundation for how to use tracing within Teleport so that more areas of the
+  code can easily be instrumented in the future
+
+### Non-Goals
+
+* Adding tracing to the entire codebase all at once
+* Replace existing logging, metrics
```

Comment:
As a side note on this, I largely agree with avoiding changing anything re: metrics. My recent experiences with Otel Tracing have been extremely positive, but the Metrics API is still incredibly immature and making a lot of breaking changes, and I've also had a few issues/bugs with their Metrics SDK. We can probably safely revisit the metrics side and Otel in six to twelve months time.